Utils for binning of spectra.

findesgh opened the pull request:
Re-implementation of the binning utilities of [pyBASS](https://gitlab.mpcdf.mpg.de/mglatzle/pybpass) using numba. I haven't quite managed to achieve the same performance as that of the original Fortran implementation, but this should not be a problem even if binning millions of spectra.

maxbriel committed:
Created CSP archistecture and basic sfh class

The complex stellar population subpackage is meant to combine bpass 
variables with a stelalr formation history to generate more complex 
populations. To this end, we're implementing a stellar formation history 
class. The basic SFH class for custom SFH is implemented using scipy 
splines. This basic class is also tested.

maxbriel committed:
Added variables to constants

maxbriel committed:
implements utility functions & new test files

maxbriel committed:
Implements event rate calculation over time

maxbriel committed:
BPASS data_files for testing file_loading

maxbriel committed:
Adds `at_time` function to the event rate class

This method allows for the calculation of the event rate at a given lookback time. It uses the BPASS binning for this. 

This method is much faster, but suffers from a lack of accuracy.

maxbriel committed:
Adds docstrings and extra tests

maxbriel committed:
Implements calculate spectrum at_time and over_time functions

Both functions are implemented for the BPASS spectra. 
However, the method is similar to the BPASS event rate calculation, which is sub optimal for the spectra calculation. It takes an extremely long time to do the calculation. A for loop of 100000 iterations is present, because each wavelength is seen as it's own "event type". It should be possible to remove this by implementing a function to take the separate wavelengths as a single unit. It would be interesting so see the influence of putting the for loop within a numba function, but thinking about restructuring the function is probably better. 

Including a "caching" of the BPASS spectra into a pickled DataFrame. 



Additional test files are required for this to run.

maxbriel committed:
Add normalisation function for the BPASS spectrum

HeloiseS committed:
fixed hrd plot

HeloiseS committed:
adding csp folder

HeloiseS committed:
fixed merge conflict

HeloiseS committed:
minor changes

maxbriel committed:
Moved tests to CSP folder + fix in function name _type_check

HeloiseS committed:
added a hokitype error

HeloiseS committed:
bringing it latest max update

HeloiseS committed:
Added some code review for Max and minor stylistic changes

HeloiseS committed:
Added some code review for Max and minor stylistic changes

maxbriel committed:
Updated the tests

maxbriel committed:
Merge remote-tracking branch 'origin/csp' into csp

HeloiseS committed:
bla

maxbriel committed:
Adjust tests and docstrings

maxbriel committed:
Adjust eventrate functions to take lists of functions

Instead of requiring a scipy interpolated spline as input. These functions now take a list of functions as input for the stellar formation and metallicity history. 

This adjustment has been made by using `numpy.interp` instead of scipy.interpolate. Furthermore, instead of scipy's spline integration `numpy.trapz` is now used to calculate the mass per bin.

HeloiseS committed:
fix merge comflict

maxbriel committed:
Disabled spectra test

HeloiseS committed:
Merge update from max 'csp' of https://github.com/HeloiseS/hoki into csp

HeloiseS committed:
Added parametric star formation histories to SFH and refactored some variable names

HeloiseS committed:
fixing CI

maxbriel committed:
Auto stash before merge of "csp" and "official_hoki/csp"

maxbriel committed:
Merge remote-tracking branch 'official_hoki/csp' into csp

maxbriel committed:
upload files for remote work

maxbriel committed:
Improve testing and functions

- Adds BPASS IMFS to the constants
- Adds check to BPASS IMFS
- Adds test for `data_compiler.SpectraData`
- Reduces data usage for rate tests

maxbriel committed:
Add SpectraCompiler test

Uses `unittest.mock` to allow for a smaller DataFrame to be used.

maxbriel committed:
Adds extensive tests for the CSP package

maxbriel committed:
Update SpectraCompiler

The compiler now outputs the BPASS spectra as a 3D `numpy.ndarray` this minimises storage and allows for faster and less memory hungry operations than the MultiIndex pandas DataFrame. 

The tests are updated accordingly and using unittest.mock a single spectra file is use. (This could be simulated too)

maxbriel committed:
Documentation update + SFH input

The documentation now shows the size and shape of each input and output variable. 

The `_over_time` and `_at_time` for `CSPSpectra` and `CSPEventRate` are now able to take the `SFH` object as an input.

maxbriel committed:
Removed CSP from utils due to cyclical import

The CSP class in `utils` was importing the SFH class, which was importing the `utils` module. Therefore, failing importing. Thus, the CSP class has been moved to a separate file

maxbriel committed:
Updated setup.cfg to include csp_test_data subfolder

maxbriel committed:
Removes leftover print statement + prints loading precompiled failed

maxbriel committed:
Correct metallicity mid-points

HeloiseS committed:
Added the parametric star formation histories as public functions

maxbriel committed:
at_time new default sample rate + load functions moved to hoki.load

at_time now uses 1000 as the default sample rate. 
The BPASS bins can still be used is a negative number is given. 
The `load_spectra` and `load_rates` have been moved to `hoki.load` and renamed to `all_spectra` and `all_rates`, respectively.
tests have been adjusted accordingly

maxbriel committed:
Fixed typo

maxbriel committed:
Merge remote-tracking branch 'origin/csp' into csp

maxbriel committed:
Removes cyclical import

maxbriel committed:
Adjusted tests that use SpectraCompiler

maxbriel committed:
Getting tox to work

maxbriel committed:
private optimised mass_per_time_bin calculation

Implements 2 new functions:
- A custom numba trapezoidal function
- A optimised mass_per_bin function for non-function-based data

maxbriel committed:
Refactored Event Rate calculation from array data

The original event rate calculation took functions as in input. This becomes cumbersome when you have data at specific time values. This calculation is done by taking a 2D matrix with time points on one axis and BPASS metallicities on the other. The values of the matrix describes the SFR at that moment in time. 

The calculate_rate_over_time function has been adjusted to use this new function.

maxbriel committed:
Removed dependency and added more doc information

Commit:
Utils for binning of spectra.

findesgh committed:
BUG: pass on kwarg to spectra compiler.

findesgh committed:
Merge branch 'csp' of https://github.com/HeloiseS/hoki into csp

Commit:
Typo.

maxbriel committed:
Merge pull request #58 from findesgh/csp

Small bugfix in `all_spectra`
The binary `kwarg` was not passed on to `SpectraCompiler` from `load.all_spectra`.

maxbriel committed:
Fix SpectraCompiler

Commit:
BUG: use np instead of pd in SpectraCompiler.

Commit:
BUG: fix index and make test pass.

Commit:
BUG: fix another regression bug.

Commit:
Merge branch 'csp' of https://github.com/HeloiseS/hoki into csp_fix2

Commit:
Fix further regression bug.

maxbriel committed:
Clean up loading module

maxbriel committed:
Updates mass per bin calculation for vectorized funciton input

mass_per_bin now takes a normal python callable and a vectorized function as input. 
The latter boosts performance significantly

maxbriel committed:
Renames CSPEventRate functions and adds grid functions

- CSPEventRate functions are now called:
    - calculate_rate_over_time -> over_time
    - calculate_rate_at_time -> at_time
new functions: 
    - grid_over_time: calculates the rates from a 2D SFH (per BPASS metallicity) from  time_points to time bins
     - grid_at_time: calculate the event rate at a specific moment in lookback time from a 2D SFH grid

maxbriel committed:
np.zeros to np.empty where possible

maxbriel committed:
Merge pull request #60 from findesgh/csp_fix2

* BUG: use np instead of pd in SpectraCompiler.
* BUG: fix index and make test pass.
* BUG: fix another regression bug.
* Fix further regression bug.

Authored-by: Martin Glatzle <mglatzle@mpa-garching.mpg.de>

maxbriel committed:
Merge remote-tracking branch 'origin/csp' into csp

maxbriel committed:
Fixed incorrect test of at_time and over_time

maxbriel committed:
Set default sample rate for at_time

Also includes some docstring updates

maxbriel committed:
Adds `cache` parameter to over_time_spectrum

maxbriel committed:
Adds grid functions for the spectra calculation

CSPSpectra now has the following functions: 
Public: 
- at_time: at time with function input
- over_time: over lb time with function input
- grid_at_time: at time with SFH grid input
- grid_over_time: over lb time with SFH grid input
Private: 
- grid_rate_calculator_at_time
- grid_rate_calculator_over_time

The grid_rate_calculators_* are numba functions that  have the grid_* around them as wrappers for the input and output

findesgh committed:
Fix limit checks on bin_edges.

maxbriel committed:
Adds return_time_edges to grid_over_time & bug fix sfh vectorisation

maxbriel committed:
Bug fix in nr_time_bins (grid_over_time functions)

Instead of N bins being created, N-1 bin was being created.

maxbriel committed:
Docs string cleanup

maxbriel committed:
Merge branch 'dev1.6' into csp

maxbriel committed:
Solve cyclical import error

maxbriel committed:
Fix tox and remove corner plot

HeloiseS commented on hoki/utils/binning.py:
```diff
@@ -0,0 +1,180 @@
+from numba import jit
+
+
+def bin_spectra(wave, spectra, bins, edges=False):
```
refactor `wave` to `wl` instead, it's just a convention I tend to use in all my codes and it's better to keep it consistent throughout hoki.

findesgh replied:
Fixed by c96e92e6fa8f54883af7f07875494710c0c45dd9.

HeloiseS commented on hoki/utils/binning.py:
```diff
@@ -0,0 +1,180 @@
+    spectra : `numpy.ndarray` (N, N_wave)
+        The SEDs to resample given as L_lambda [Energy/Time/Wavelength] or L_nu
+        [Energy/Time/Frequency] in accordance with `wave`.
+    bins : `numpy.ndarray` (N_bins,)
```
A little clarification needed in this definition. 

Just reading this it's not clear whether i can give the number of *old bins* I want in *new bins* (e.g. combining 3 bins into one) or if I have to give central wavelength (or edges, that latter part is clear).  If the option is central wavelength or edge wavelengths I suggest a minor refactor:

instead of having a `bins` and `edges`  parameter, have a `wl_center` and `wl_edges` parameters - only one should be filled and you an check for that. Default them to None (`wl_center=None` and `wl_edges=None`) and if both remain non on execution print a custom error message with a 'HOKI DEBUGGER:" explaining to the use they need to provide at least one of these.

findesgh replied:
Fair point. I've overhauled how the bins work, bringing it more in line with what one knows from `numpy` and also updated the docstrings. Should hopefully be much clearer now.

HeloiseS commented on hoki/utils/binning.py:
```diff
@@ -0,0 +1,180 @@
+
+def bin_spectra(wave, spectra, bins, edges=False):
+    """
+    Bin spectra conserving luminosity.
```
that is key - and very interesting to me. I have a (personal - no yet in hoki) binning algorithm very different from yours, because it is designed to bin spectra in **counts** (and calculate associated errors). 

I am wondering if we could call this function something more specific to indicate this is specifically designed to handle luminosities (and other non-Poissonian quantities). this way when I finally add my binning algo there won't be a clash in naming.

findesgh replied:
Hmm. I'm sure there is a better name. But I'm not clear on how your application is different or rather how it works in the first place so I'm struggling to come up with something of the cuff. I have no problem with renaming things, of course, but I think I'd first need to understand what the difference actually is.

HeloiseS replied:
fair - I would recommend we then call it `bin_luminosity`, mine would then be called `bin_counts` :)
Does that work for you?

findesgh replied:
Done.

HeloiseS reviewed (changes requested):
Excellent work  and  must apologise for the time it took me to get to this: I was taking a course on deep learning and then had to take sick leave to treat my HVS. 

I've added a few comments to the files and I have one main questions: is there a reason you added these function to the `utils` module and not the `spec` module which specifically has to do with spectroscopy?

findesgh commented:
No worries, thanks for taking a look!

Regarding the location in `utils`: I used `csp` as a base branch for this since I need the spectra loading functionality for #46. And the `spec` module has not yet been merged into `csp`. In principle, though, feel free to move the code to where you see fit :-).

Will address the other issues as time permits.

findesgh committed:
REFACTOR: wave to wl.

findesgh committed:
DOC: update to docstrings.

findesgh committed:
Overhaul how bins work.

maxbriel committed:
File check, combined data_compiler tests, renames func+variables

renamed variables:
- BPASS_METALLICITIES. now with `z`
renamed functions:
- all_rates ->  rates_all_z
- all_spectra -> spectra_all_z

findesgh committed:
Merge branch 'csp' into rebin

findesgh committed:
Merge branch 'dev1.6' into rebin

findesgh commented:
Merged `dev1.6` into this and chose it as base branch as well, as that makes more sense, I suppose. Should I move the functionality to `hoki/spec.py`?

HeloiseS commented:
> Merged `dev1.6` into this and chose it as base branch as well, as that makes more sense, I suppose. Should I move the functionality to `hoki/spec.py`?

great, that makes sense :)

Commit:
REFACTOR: bin_spectra to bin_luminosity.

Commit:
REFACTOR: move binning utils to spec.
